api/client/charms/localcharmputters.go

```diff
@@ -0,0 +1,165 @@
+// hosted by the controller
+func newHTTPPutter(api base.APICaller) (CharmPutter, error) {
+	// The returned httpClient sets the base url to /model/:modeluuid if it can.
+	apiHTTPClient, err := api.HTTPClient()
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot retrieve http client from the api connection")
+	}
+	return &httpPutter{
+		httpClient: apiHTTPClient,
+	}, nil
+}
+
+func (h *httpPutter) PutCharm(ctx context.Context, _, _, curlStr string, body io.Reader) (string, error) {
+	curl, err := charm.ParseURL(curlStr)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	args := url.Values{}
+	args.Add("series", curl.Series)
+	args.Add("schema", curl.Schema)
+	args.Add("revision", strconv.Itoa(curl.Revision))
+	apiURI := url.URL{Path: "/charms", RawQuery: args.Encode()}
+
+	var resp params.CharmsResponse
+	req, err := http.NewRequest("POST", apiURI.String(), body)
+	if err != nil {
+		return "", errors.Annotate(err, "cannot create upload request")
+	}
+	req.Header.Set("Content-Type", "application/zip")
+
+	if err := h.httpClient.Do(ctx, req, &resp); err != nil {
+		return "", errors.Trace(err)
+	}
+	return resp.CharmURL, nil
+}
+
+// s3Putter uploads local charm blobs to the controller via the
+// s3-compatible api endpoint "model-:modeluuid/charms/:object"
+type s3Putter struct {
+	httpClient *httprequest.Client
+}
+
+// newS3Putter creates an s3Putter, which uploads local charm blobs
+// to the controller via the s3-compatible api endpoint
+// "model-:modeluuid/charms/:object"
+func newS3Putter(api base.APICaller) (CharmPutter, error) {
+	apiHTTPClient, err := api.RootHTTPClient()
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot retrieve http client from the api connection")
+	}
+	return &s3Putter{
+		httpClient: apiHTTPClient,
+	}, nil
+}
+
+func (h *s3Putter) PutCharm(ctx context.Context, modelUUID, charmRef, curl string, body io.Reader) (string, error) {
+	apiURI := url.URL{Path: fmt.Sprintf("/model-%s/charms/%s", modelUUID, charmRef)}
+
+	resp := &http.Response{}
+	req, err := http.NewRequest("PUT", apiURI.String(), body)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	req.Header.Set("Content-Type", "application/zip")
+	req.Header.Set("Juju-Curl", curl)
+
+	if err := h.httpClient.Do(ctx, req, &resp); err != nil {
+		return "", errors.Trace(err)
+	}
+
+	return resp.Header.Get("Juju-Curl"), nil
+}
+
+// fallbackPutter iterates over a number of sub-putters, attempting to upload a charm
+// blob until one is successful. If a putter fails, before falling back to the next one
+// we check is the error is 'fallback-able'. It doesn't make sense to fallback on certain
+// error types such as unauthorised
+type fallbackPutter struct {
+	putters []CharmPutter
+}
+
+// newFallbackPutter creates a fallbackPutter with at least 2 sub-putters
+func newFallbackPutter(primaryPutter, secondaryPutter CharmPutter, teritaryPutters ...CharmPutter) CharmPutter {
```

Comment:
As we discussed.

```suggestion
func newFallbackPutter(putters ...CharmPutter) CharmPutter {
```

Fail with programming error if there isn't at least one putter.